keanu-project/src/main/java/io/improbable/keanu/distributions/tensors/continuous/TensorSmoothUniform.java

```diff
@@ -0,0 +1,119 @@
+        final DoubleTensor rScaled = r1.times(bodyWidth.plus(shoulderWidth)).plusInPlace(xMin.minus(shoulderWidth.div(2)));
+        final DoubleTensor bodyHeight = bodyHeight(shoulderWidth, bodyWidth);
+
+        final DoubleTensor firstConditional = rScaled.getGreaterThanOrEqualToMask(xMin);
+        firstConditional.timesInPlace(rScaled.getLessThanOrEqualToMask(xMax));
+        final DoubleTensor inverseFirstConditional = DoubleTensor.ones(firstConditional.getShape()).minusInPlace(firstConditional);
+
+        final DoubleTensor secondConditional = rScaled.getLessThanMask(xMin);
+        DoubleTensor spillOnToShoulder = xMin.minus(rScaled);
+        DoubleTensor shoulderX = shoulderWidth.minus(spillOnToShoulder);
+        DoubleTensor shoulderDensity = shoulder(shoulderWidth, bodyWidth, shoulderX);
+        DoubleTensor acceptProbability = shoulderDensity.div(bodyHeight);
+
+        final DoubleTensor secondConditionalNestedTrue = secondConditional.times(r2.getLessThanOrEqualToMask(acceptProbability));
+        final DoubleTensor secondConditionalNestedFalse = secondConditional.times(r2.getGreaterThanMask(acceptProbability));
+        final DoubleTensor secondConditionalNestedFalseResult = xMin.minus(shoulderWidth).plus(spillOnToShoulder);
+
+        final DoubleTensor secondConditionalFalse = rScaled.getGreaterThanOrEqualToMask(xMin);
+        spillOnToShoulder = rScaled.minus(xMax);
+        shoulderX = shoulderWidth.minus(spillOnToShoulder);
+        shoulderDensity = shoulder(shoulderWidth, bodyWidth, shoulderX);
+        acceptProbability = shoulderDensity.divInPlace(bodyHeight);
+
+        final DoubleTensor secondConditionalFalseNestedTrue = secondConditionalFalse.times(r2.getLessThanOrEqualToMask(acceptProbability));
+        final DoubleTensor secondConditionalFalseNestedFalse = secondConditionalFalse.times(r2.getGreaterThanMask(acceptProbability));
+        final DoubleTensor secondConditionalFalseNestedFalseResult = xMax.plus(shoulderWidth).minusInPlace(spillOnToShoulder);
+
+        return firstConditional.times(rScaled).
+            plus(inverseFirstConditional.times(secondConditionalNestedTrue).times(rScaled)).
+            plus(inverseFirstConditional.times(secondConditionalNestedFalse).times(secondConditionalNestedFalseResult)).
+            plus(inverseFirstConditional.times(secondConditionalFalseNestedTrue).times(rScaled)).
+            plus(inverseFirstConditional.times(secondConditionalFalseNestedFalse).times(secondConditionalFalseNestedFalseResult));
+    }
+
+    public static DoubleTensor pdf(DoubleTensor xMin, DoubleTensor xMax, DoubleTensor shoulderWidth, DoubleTensor x) {
+
+        final DoubleTensor bodyWidth = xMax.minus(xMin);
+        final DoubleTensor rightCutoff = xMax.plus(shoulderWidth);
+        final DoubleTensor leftCutoff = xMin.minus(shoulderWidth);
+
+        final DoubleTensor firstConditional = x.getGreaterThanOrEqualToMask(xMin);
```

Comment:
these inPlace ops should reassign just in case the inPlace can't be done in the same tensor.

firstConditional = firstConditional....InPlace(...)
secondConditional = 
thirdConditional =

Reply:
nice, changed